cell_classification/predict.py

```diff
@@ -0,0 +1,236 @@
+
+    # save params to toml file
+    with open(os.path.join(params["model_dir"], "params.toml"), "w") as f:
+        toml.dump(params, f)
+    return single_example_list, params
+
+
+def calc_roc(pred_list, gt_key="marker_activity_mask", pred_key="prediction"):
+    """Calculate ROC curve
+    Args:
+        pred_list (list):
+            list of samples with predictions
+        gt_key (str):
+            key for ground truth labels
+        pred_key (str):
+            key for predictions
+    Returns:
+        roc (dict):
+            dictionary containing ROC curve data
+    """
+    roc = {"fpr": [], "tpr": [], "thresholds": [], "auc": []}
+    for sample in pred_list:
+        if sample[gt_key].max() == 0:
+            continue
+        fpr, tpr, thresholds = roc_curve(sample[gt_key].flatten(), sample[pred_key].flatten())
+        roc["fpr"].append(fpr)
+        roc["tpr"].append(tpr)
+        roc["thresholds"].append(thresholds)
+        roc["auc"].append(auc(fpr, tpr))
+    return roc
+
+
+def calc_metrics(pred_list, gt_key="marker_activity_mask", pred_key="prediction"):
+    """Calculate metrics
+    Args:
+        pred_list (list):
+            list of samples with predictions
+        gt_key (str):
+            key of ground truth in pred_list
+        pred_key (str):
+            key of prediction in pred_list
+    Returns:
+        avg_metrics (dict):
+            dictionary containing metrics averaged over all samples
+    """
+    metrics_dict = {
+        "accuracy": [],
+        "precision": [],
+        "recall": [],
+        "f1_score": [],
+        "tp": [],
+        "tn": [],
+        "fp": [],
+        "fn": [],
+    }
+
+    def _calc_metrics(threshold):
+        """Calculate metrics for a given threshold"""
+        metrics = deepcopy(metrics_dict)
+        for sample in pred_list:
+            if sample[gt_key].max() == 0:
+                continue
+            tn, fp, fn, tp = confusion_matrix(
+                y_true=sample[gt_key].flatten(),
+                y_pred=(sample[pred_key].flatten() >= threshold).astype(int),
+            ).ravel()
+            metrics["tp"].append(tp)
+            metrics["tn"].append(tn)
+            metrics["fp"].append(fp)
+            metrics["fn"].append(fn)
+            metrics["accuracy"].append((tp + tn) / (tp + tn + fp + fn))
+            metrics["precision"].append(tp / (tp + fp + 1e-8))
+            metrics["recall"].append(tp / (tp + fn + 1e-8))
+            metrics["f1_score"].append(2 * tp / (2 * tp + fp + fn))
+        metrics["threshold"] = threshold
+        for key in ["dataset", "imaging_platform", "marker"]:
+            metrics[key] = sample[key]
+        return metrics
+
+    thresholds = np.linspace(0, 1, 101)
+    metric_list = Parallel(n_jobs=8)(delayed(_calc_metrics)(i) for i in thresholds)
+    avg_metrics = deepcopy(metrics_dict)
+    for key in ["dataset", "imaging_platform", "marker", "threshold"]:
+        avg_metrics[key] = []
```

Comment:
In general this file has pretty dense code without many comments

Reply:
I added more comments and to make it more understandable.